test/performance/io/format_sam_benchmark.cpp

```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    return mutated_sequence;
+}
+
+// ============================================================================
+// generate sam file from randomly generated sequence and store in file_dict
+// ============================================================================
+std::map<std::size_t, std::string> file_dict;
+std::string const default_query_id = "query_";
+std::string reference_id = "reference_id";
+
+static std::string get_file (std::size_t n_queries)
+{
+    if (file_dict.find(n_queries) == file_dict.end())
+    {
+        std::string file{};
+
+        // generate sequences
+        auto reference = test::generate_sequence<dna4>(reference_length, 0, default_seed);
+        std::vector<std::vector<dna4>> queries;
+
+        size_t seed = default_seed;
+        for (size_t i = 0; i < n_queries; ++i, ++seed)
+        {
+            auto query = mutate_sequence<dna4>(reference, seed);
+            queries.push_back(query);
+        }
+
+        // align
+        auto config = align_cfg::mode{global_alignment} |
+                      align_cfg::scoring{nucleotide_scoring_scheme{match_score{4}, mismatch_score{-2}}} |
+                      align_cfg::gap{gap_scheme{gap_score{-4}}} |
+                      align_cfg::aligned_ends{free_ends_all} |
+                      align_cfg::result{with_alignment};
+
+        std::stringstream sam_stream;
+        alignment_file_format_sam format{};
+
+        alignment_file_output sam_out{sam_stream,
+                                      format,
+                                      fields< field::SEQ,
+                                              field::ID,
+                                              field::REF_ID,
+                                              field::REF_SEQ,
+                                              field::ALIGNMENT,
+                                              field::MAPQ>{}};
+
+        size_t i = 0;
+        for(auto query : queries)
+        {
+            auto &result = *(align_pairwise(std::tie(reference, query), config).begin());
+            auto aligned_sequence = result.alignment();
+            size_t map_qual = 60u + result.score();
+            std::string current_query_id = default_query_id + std::to_string(i);
+            ++i;
+
+            sam_out.emplace_back(query,
+                                 current_query_id,
+                                 reference_id,
+                                 reference,
+                                 aligned_sequence,
+                                 map_qual);
+
+        }
+
+        // convert to string
+        file = sam_stream.str();
+
+        file_dict[n_queries] = file;
+    }
+
+    return file_dict[n_queries];
+}
+
+// ============================================================================
+// read as many sam files as possible
+// ============================================================================
+void sam_read_from_stream (benchmark::State & state)
```

Comment:
please exchange the format reading by directly reading from the file. this is easier and will model the reality better :)

```cpp
void sam_file_read_from_stream (benchmark::State & state)
{
    size_t n_queries = state.range(0);
    std::string sam_file = get_file(n_queries);

    std::istringstream istream{sam_file};

    for (auto _ : state)
    {
        // refill stream (same constant for seqan2 benchmark)
        istream.clear();
        istream.seekg(0, std::ios::beg);

        alignment_file_input fin{istream, format_sam{}};

        // read all records and store in internal buffer
        auto it = fin.begin();
        while (it != fin.end())
            ++it;
    }
}
```